### DISCUSSION TITLE
docs/current/tcpservice

### DISCUSSION THREAD

**Discussion by giscus** [RRQM/RRQMSocket#40] [Announcements]
# docs/current/tcpservice

定义

https://touchsocket.net/docs/current/tcpservice

**Comment by RRQM** [comment]
测试评论

**Comment by YAOZZJ** [comment]
我想实现一个功能被是同步堵塞接收,请问需要怎么弄呀?就是server 发送了数据后,直接在上下文等待client返回数据

**Reply by RRQM** [reply]
你可以使用WaitingClient来实现功能。

https://touchsocket.net/docs/current/waitingclient

也可以看下我们的视频课程，里面有更详细的设计思路和教学。

https://www.bilibili.com/cheese/play/ep1541514

**Comment by GuDiaoBuTan** [comment]
我用TcpService在WPF中是实现了一个采集器数据的远程监测，采集器每4500毫秒发送一次数据。运行中发现一个很奇怪的现象：程序明明没有主动关闭采集器的socket连接，但是只要监测到10分钟左右，采集器就会重新发起连接请求。但是我用网络调试助手作为服务端来长时间接收采集器数据确实正常的，这是怎么回事

**Comment by chuanjie1059** [comment]
如何一次性给所有的Client发送数据？

**Reply by RRQM** [reply]
你可以直接从Service中，直接获取当前在线的所有客户端。然后进行发送操作。

`
foreach (var client in service.Clients)
{
    await client.SendAsync("hello");
}
`

**Comment by att1008** [comment]
为什么我客户端（.net原生）如果是国外的IP连接到我的服务器（touchsocket），连接会自动断开？我看断开事件里面的message显示是客户端主动断开，但是我用客户端调试发现客户端没有主动断开并且客户端异步阻塞读取数据返回的字节数是0代表被断开？

**Comment by tstrd962-del** [comment]
Hi, Is there a solution to  tcpClient: "The remote terminal actively disconnects"?

**Comment by jieme1989** [comment]
HttpSessionClient的Protocol为websocket，Protocol.WebSocket 是ws ？？？

private bool IsClientEnable(HttpSessionClient client)
{
    try
    {
        return client != null &&
               client.Protocol == Protocol.WebSocket && //这一行不相等
               client.WebSocket != null &&
               client.WebSocket.Online &&
               !client.DisposedValue;
    }
    catch
    {
        return false;
    }
}
